service/hook/github/metrics.go

```diff
@@ -0,0 +1,269 @@
+package github
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/bitrise-io/bitrise-webhooks/service/hook/common"
+	hookCommon "github.com/bitrise-io/bitrise-webhooks/service/hook/common"
+	"github.com/google/go-github/v54/github"
+)
+
+// GatherMetrics ...
+func (hp HookProvider) GatherMetrics(r *http.Request, appSlug string) (common.Metrics, error) {
+	payload, err := github.ValidatePayload(r, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	webhookType := github.WebHookType(r)
+
+	event, err := github.ParseWebHook(webhookType, payload)
+	if err != nil {
+		return nil, err
+	}
+
+	currentTime := hp.timeProvider.CurrentTime()
+	return hp.gatherMetrics(event, webhookType, appSlug, currentTime), nil
+}
+
+func (hp HookProvider) gatherMetrics(event interface{}, webhookType, appSlug string, currentTime time.Time) common.Metrics {
```

Comment:
event could be more strongly typed

Reply:
What about using this interface for MetricsProvider:

```
type MetricsResult struct {
	PushMetrics               *PushMetrics
	PullRequestMetrics        *PullRequestMetrics
	PullRequestCommentMetrics *PullRequestCommentMetrics
}

// MetricsProvider ...
type MetricsProvider interface {
	GatherMetrics(r *http.Request, appSlug string) (MetricsResult, error)
}
```

This way the code better signals that one of the 3 Metrics can be returned.

This would be used like this:
```
	res, err = metricsProvider.GatherMetrics(r, appSlug)
	switch {
	case res.PushMetrics != nil:
		webhookMetrics = res.PushMetrics
	case res.PullRequestMetrics != nil:
		webhookMetrics = res.PushMetrics
	case res.PullRequestCommentMetrics != nil:
		webhookMetrics = res.PullRequestCommentMetrics
	}
```

Reply:
turns out that the idea I had is not yet supported by go :( https://stackoverflow.com/questions/72024835/go-generics-why-doesnt-the-switch-case-aany-match-aint


I tried to avoid the interface{} type and I'm not sure if the 3 pointers with only 1 value is much safer, so let's leave it as it is and improve it later when go has more support for generics.